src/GNDStk/HDF5.hpp

```diff
@@ -0,0 +1,126 @@
+
+// -----------------------------------------------------------------------------
+// HDF5
+// Wraps HighFive::File
+// -----------------------------------------------------------------------------
+
+class HDF5 {
+public:
+
+   // data
+   HighFive::File *file = nullptr;
+   std::string filename = "";
+   mutable bool temporary = false;
+
+   // file modes
+   static inline const auto modeRead =
+      HighFive::File::ReadOnly;
+   static inline const auto modeWrite =
+      HighFive::File::ReadWrite |
+      HighFive::File::Create |
+      HighFive::File::Truncate;
+
+   // temporaryName()
+   static std::string temporaryName()
+   {
+      // Generate a file name that's supposed to be suitable for safely creating
+      // a temporary file. L_tmpnam and tmpnam() are from <cstdio>. In case
+      // anyone wonders, we don't need a +1 in the [L_tmpnam]. :-)
+      static char buffer[L_tmpnam];
+
+      while (true) {
+         const std::string name = std::string(tmpnam(buffer)) + ".h5";
```

Comment:
I am getting tmpnam is deprecated warning on clang 12. Due to us handling every warning as an error, it does not compile onn clang 12. We should replace this asap.

Reply:
This was long since replaced with better code in later PRs, which no longer get that warning.

Reply:
Good. I switched to g++ 9.3 on my laptop so I can actually compile it now.

Reply:
By the way, it's worth explaining why I deal with temporary files when I'm working with HDF5.

For ensuring consistency of behavior, and making various things throughout GNDStk (reads, writes; XML, JSON, and HDF5; streams and files; convert functions; etc.) work smoothly and consistently with one another, I have lightweight wrappers around pugixml's data structure for XML, nlohmann json's data structure for JSON, and HighFive's data structure for HDF5. (Which, itself, actually relies on a large and complex underlying C-based library written by the official HDF5 people.)

When XML is read by pugixml, the whole file is put into pugxmli's internal data structure. So far so good.

When JSON is read by nlohmann json, the whole file is put into nlohmann json's internal data structure. So far so good.

Oops...HighFive (probably because of the underlying C-based library) is different!! Its data structure is small, and mostly serves as essentially a *file* *descriptor* that points to a file on disk! HDF5 data aren't actually stored in memory. :-(

That throws a monkey wrench into the whole system. So, of course, wanting GNDStk to be as awesome a code as possible, my HDF5 wrapper un-monkey-wrenches things.

Imagine a simple implementation that was unaware of what the underlying hdf5-related libraries actually do. You might "load" an HDF5 file into an object of my HDF5 class, then copy that to a second HDF5 object that you wish to modify. Unfortunately, you'd basically be dealing with a shallow copy - two things that amount to file descriptors to the same underlying file. You can imagine various bad things that would happen.

So, I wrote all of my HDF5 capabilities to be fully aware of the issue. Just say - to look at a simple case - that you do what I describe above: read into an object of class HDF5, then copy to another. For proper semantics with no surprises, GNDStk's copy capability *duplicates* the original file (for which the source HDF5 object has a file descriptor) into a temporary file. (Hence the need to generate temporary file names!) Then, the copy's destination HDF5 is set up to have a file descriptor to the temporary file, not to the original file.

I carefully wrote all the behavior of all of GNDStk's HDF5 I/O to deal with this, so that a high-level user will just see an interface that looks exactly the same as the interfaces for XML and JSON. This was a tricky task, but one that I think the best GNDS code in the world deserves!!

Be aware, then, that if you happen to look in detail at everything I did for HDF5, you'll see various *apparently* funny gymnastics and convolutednesses (if that's a word) that might seem bizarre and unnecessary, unless you understand that I'm dealing with underlying code that doesn't (or only partially) has actual HDF5 data structures in memory.